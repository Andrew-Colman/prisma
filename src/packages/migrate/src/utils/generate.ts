const childProcess = require('child_process')
const { promisify } = require('util')

const exec = promisify(childProcess.exec)

export async function generate() {
  const localPath = getLocalPackagePath()
  console.log(localPath)
  try {
    if (localPath) {
      await run('node', [localPath, 'generate'])
    }
  } catch (e) {
    console.error(e)
  }
  return
}

function getLocalPackagePath() {
  let packagePath
  try {
    packagePath = require.resolve('@prisma/cli/package.json')
  } catch (e) {
    return null
  }

  if (packagePath) {
    return require.resolve('@prisma/cli')
  }

  return null
}

function run(cmd, params) {
  const child = childProcess.spawn(cmd, params, {
    stdio: ['pipe', 'inherit', 'inherit'],
  })

  return new Promise((resolve, reject) => {
    child.on('close', () => {
      resolve()
    })
    child.on('exit', (code) => {
      if (code === 0) {
        resolve()
      } else {
        reject(code)
      }
    })
    child.on('error', () => {
      reject()
    })
  })
}
