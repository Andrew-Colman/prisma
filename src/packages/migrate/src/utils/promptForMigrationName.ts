import { prompt } from 'prompts'
import { isCi } from '@prisma/sdk'

export async function getMigrationName(name?: string): Promise<string> {
  if (name) {
    return name
  } else if (isCi()) {
    return ''
  }

  const response = await prompt({
    type: 'text',
    name: 'name',
    message: `Name of migration`,
  })
  return response.name || ''
}
