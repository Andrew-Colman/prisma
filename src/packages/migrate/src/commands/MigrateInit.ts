import {
  arg,
  Command,
  format,
  getCommandWithExecutor,
  getSchemaPath,
  HelpError,
  isError,
} from '@prisma/sdk'
import chalk from 'chalk'
import path from 'path'
import { Migrate } from '../Migrate'
import { ensureDatabaseExists } from '../utils/ensureDatabaseExists'
import { ExperimentalFlagError } from '../utils/experimental'

export class MigrateInit implements Command {
  public static new(): MigrateInit {
    return new MigrateInit()
  }

  private static help = format(`
    Initialize migrations

    ${chalk.bold.yellow('WARNING')} ${chalk.bold(
    "Prisma's migration functionality is currently in an experimental state.",
  )}
    ${chalk.dim(
      'When using any of the commands below you need to explicitly opt-in via the --experimental flag.',
    )}

    ${chalk.bold('Usage')}

      ${chalk.dim('$')} prisma migrate init --experimental

    ${chalk.bold('Options')}

      -h, --help       Displays this help message

  `)

  public async parse(argv: string[]): Promise<string | Error> {
    const args = arg(argv, {
      '--help': Boolean,
      '-h': '--help',
      '--experimental': Boolean,
      '--schema': String,
      '--telemetry-information': String,
    })

    if (isError(args)) {
      return this.help(args.message)
    }

    if (args['--help']) {
      return this.help()
    }

    if (!args['--experimental']) {
      throw new ExperimentalFlagError()
    }

    const schemaPath = await getSchemaPath(args['--schema'])

    if (!schemaPath) {
      throw new Error(
        `Could not find a ${chalk.bold(
          'schema.prisma',
        )} file that is required for this command.\nYou can either provide it with ${chalk.greenBright(
          '--schema',
        )}, set it as \`prisma.schema\` in your package.json or put it into the default location ${chalk.greenBright(
          './prisma/schema.prisma',
        )} https://pris.ly/d/prisma-schema-location`,
      )
    }

    console.info(
      chalk.dim(
        `Prisma Schema loaded from ${path.relative(process.cwd(), schemaPath)}`,
      ),
    )

    // Automatically create the database if it doesn't exist
    await ensureDatabaseExists('create', true, schemaPath)

    const migrate = new Migrate(schemaPath)
    await migrate.initialize()
    // after engine init is implemented
    // migrate.stop()

    return `\nInitialization complete.\n\nRun ${chalk.greenBright(
      getCommandWithExecutor('prisma migrate --experimental'),
    )} to create your first migration\n`
  }

  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(
        `\n${chalk.bold.red(`!`)} ${error}\n${MigrateInit.help}`,
      )
    }
    return MigrateInit.help
  }
}
