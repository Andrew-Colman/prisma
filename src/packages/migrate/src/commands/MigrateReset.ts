import {
  arg,
  Command,
  format,
  getCommandWithExecutor,
  getSchemaPath,
  HelpError,
  isError,
  isCi,
} from '@prisma/sdk'
import chalk from 'chalk'
import path from 'path'
import prompt from 'prompts'
import { Migrate } from '../Migrate'
import { ExperimentalFlagError } from '../utils/experimental'
import { printFilesFromMigrationIds } from '../utils/printFiles'

export class MigrateReset implements Command {
  public static new(): MigrateReset {
    return new MigrateReset()
  }

  private static help = format(`
    Reset your database and reapply migrations

    ${chalk.bold.yellow('WARNING')} ${chalk.bold(
    "Prisma's migration functionality is currently in an experimental state.",
  )}
    ${chalk.dim(
      'When using any of the commands below you need to explicitly opt-in via the --experimental flag.',
    )}

    ${chalk.bold('Usage')}

      ${chalk.dim('$')} prisma migrate reset --experimental

    ${chalk.bold('Options')}

      -h, --help       Displays this help message
      -f, --force      Skip the confirmation prompt

  `)

  public async parse(argv: string[]): Promise<string | Error> {
    const args = arg(argv, {
      '--help': Boolean,
      '-h': '--help',
      '--force': Boolean,
      '-f': '--force',
      '--experimental': Boolean,
      '--schema': String,
      '--telemetry-information': String,
    })

    if (isError(args)) {
      return this.help(args.message)
    }

    if (args['--help']) {
      return this.help()
    }

    if (!args['--experimental']) {
      throw new ExperimentalFlagError()
    }

    const schemaPath = await getSchemaPath(args['--schema'])

    if (!schemaPath) {
      throw new Error(
        `Could not find a ${chalk.bold(
          'schema.prisma',
        )} file that is required for this command.\nYou can either provide it with ${chalk.greenBright(
          '--schema',
        )}, set it as \`prisma.schema\` in your package.json or put it into the default location ${chalk.greenBright(
          './prisma/schema.prisma',
        )} https://pris.ly/d/prisma-schema-location`,
      )
    }

    console.info(
      chalk.dim(
        `Prisma Schema loaded from ${path.relative(process.cwd(), schemaPath)}`,
      ),
    )

    if (!args['--force']) {
      if (isCI()) {
        throw Error(
          `Use the --force flag to use the reset command in an unnattended environment like ${chalk.bold.greenBright(
            getCommandWithExecutor('prisma reset --force --experimental'),
          )}`,
        )
      }

      const confirmation = await prompt({
        type: 'confirm',
        name: 'value',
        message: `Are you sure you want to reset your database? ${chalk.red(
          'All data will be lost',
        )}.`,
      })

      if (!confirmation.value) {
        console.info('Reset cancelled.')
        process.exit(0)
      }
    }

    const migrate = new Migrate(schemaPath)

    await migrate.reset()

    const migrationIds = await migrate.applyOnly()

    migrate.stop()

    if (migrationIds.length === 0) {
      return `\nDatabase reset successful, Prisma Migrate didn't find unapplied migrations.\n`
    } else {
      return `\nDatabase reset successful, Prisma Migrate applied the following migration(s):\n\n${chalk.dim(
        printFilesFromMigrationIds('migrations', migrationIds, {
          'migration.sql': '',
        }),
      )}\n`
    }
  }

  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(
        `\n${chalk.bold.red(`!`)} ${error}\n${MigrateReset.help}`,
      )
    }
    return MigrateReset.help
  }
}
