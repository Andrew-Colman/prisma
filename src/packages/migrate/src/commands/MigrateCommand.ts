import {
  arg,
  Command,
  Commands,
  format,
  HelpError,
  isError,
  unknownCommand,
  getSchemaPath,
  getCommandWithExecutor,
  isCi,
  IntrospectionEngine,
} from '@prisma/sdk'
import Debug from '@prisma/debug'
import chalk from 'chalk'
import prompt from 'prompts'
import path from 'path'
import { Migrate } from '../Migrate'
import { ensureDatabaseExists } from '../utils/ensureDatabaseExists'
import {
  EarlyAcessFlagError,
  ExperimentalFlagWithNewMigrateError,
} from '../utils/flagErrors'
import { printMigrationId } from '../utils/printMigrationId'
import { printFilesFromMigrationIds } from '../utils/printFiles'
import {
  handleUnexecutableSteps,
  handleWarnings,
} from '../utils/handleEvaluateDataloss'
import { getMigrationName } from '../utils/promptForMigrationName'
import { isOldMigrate } from '../utils/detectOldMigrate'

const debug = Debug('migrate:cmd')

/**
 * Migrate command
 */
export class MigrateCommand implements Command {
  public static new(cmds?: Commands): MigrateCommand {
    return new MigrateCommand(cmds)
  }

  private static help = format(`
    ${
      process.platform === 'win32' ? '' : chalk.bold('üèãÔ∏è  ')
    }Migrate your database with confidence

    ${chalk.bold.yellow('WARNING')} ${chalk.bold(
    "Prisma's migration functionality is currently in an experimental state.",
  )}
    ${chalk.dim(
      'When using any of the commands below you need to explicitly opt-in via the --early-access-feature flag.',
    )}

    ${chalk.bold('Usage')}

      With an existing schema.prisma:
      ${chalk.dim(
        '$',
      )} prisma migrate [command] [options] --early-access-feature

      Or specify a schema:
      ${chalk.dim(
        '$',
      )} prisma migrate [command] [options] --early-access-feature --schema=./schema.prisma

    ${chalk.bold('Options')}

      -h, --help   Display this help message
         --draft   Create a draft of a migration that can be edited locally before being applied

    ${chalk.bold('Commands')}

          up      Migrate your database up
          reset   Reset your database, all data will be lost

    ${chalk.bold('Examples')}

      Create a new migration and apply it
      ${chalk.dim('$')} prisma migrate --early-access-feature

      Reset your database
      ${chalk.dim('$')} prisma migrate reset --early-access-feature

      Create a draft of a migration
      ${chalk.dim('$')} prisma migrate --draft --early-access-feature
  `)

  private argsSpec = {
    '--help': Boolean,
    '-h': '--help',
    '--name': String,
    '-n': '--name',
    '--force': Boolean,
    '-f': '--force',
    '--draft': Boolean,
    '--schema': String,
    '--skip-generate': Boolean,
    '--experimental': Boolean,
    '--early-access-feature': Boolean,
    '--telemetry-information': String,
  }

  private constructor(private readonly cmds?: Commands) {}

  public async parse(argv: string[]): Promise<string | Error> {
    const args = arg(argv, this.argsSpec)

    if (isError(args)) {
      return this.help(args.message)
    }

    if (args['--help']) {
      return this.help()
    }

    if (args['--experimental']) {
      throw new ExperimentalFlagWithNewMigrateError()
    }

    const schemaPath = await getSchemaPath(args['--schema'])
    if (schemaPath) {
      const migrationDirPath = path.join(path.dirname(schemaPath), 'migrations')
      if (isOldMigrate(migrationDirPath)) {
        // Maybe add link to docs?
        throw Error(
          `The migrations folder contains migrations files from an older version of Prisma Migrate which is not compatible.
Delete the current migrations folder to continue and read the documentation for how to upgrade / baseline.`,
        )
      }
    }

    // running a subcommand
    if (args._[0] && this.cmds) {
      // check if we have that subcommand
      const cmd = this.cmds[args._[0]]
      if (cmd) {
        const argsForCmd = args['--early-access-feature']
          ? [
              ...args._.slice(1),
              `--early-access-feature=${args['--early-access-feature']}`,
            ]
          : args._.slice(1)
        return cmd.parse(argsForCmd)
      }

      return unknownCommand(MigrateCommand.help, args._[0])
    } else {
      // prisma migrate
      if (!args['--early-access-feature']) {
        throw new EarlyAcessFlagError()
      }

      return await this.migrate(argv)
    }
  }

  // All-in-One command
  public async migrate(argv: string[]): Promise<string> {
    const args = arg(argv, this.argsSpec)

    const schemaPath = await getSchemaPath(args['--schema'])

    if (!schemaPath) {
      const message = `Could not find a ${chalk.bold(
        'schema.prisma',
      )} file that is required for this command.\nYou can either provide it with ${chalk.greenBright(
        '--schema',
      )}, set it as \`prisma.schema\` in your package.json or put it into the default location ${chalk.greenBright(
        './prisma/schema.prisma',
      )} https://pris.ly/d/prisma-schema-location`

      // if no schema arg passed display error + help
      if (!args['--schema']) {
        throw this.help(message) as string
      }

      // else only display error
      throw new Error(message)
    }

    console.info(
      chalk.dim(
        `Prisma Schema loaded from ${path.relative(process.cwd(), schemaPath)}`,
      ),
    )

    // Automatically create the database if it doesn't exist
    const wasDbCreated = await ensureDatabaseExists('create', true, schemaPath)
    if (wasDbCreated) {
      console.info()
      console.info(wasDbCreated)
    }

    const migrate = new Migrate(schemaPath)

    if (args['--draft']) {
      let migrationName: undefined | string = undefined
      const getMigrationNameResult = await getMigrationName(args['--name'])
      if (getMigrationNameResult.userCancelled) {
        migrate.stop()
        return getMigrationNameResult.userCancelled
      } else {
        migrationName = getMigrationNameResult.name
      }

      const migrationId = await migrate.draft({
        name: migrationName,
      })
      migrate.stop()

      if (migrationId) {
        return `\nPrisma Migrate created a draft migration ${printMigrationId(
          migrationId,
        )}\n\nYou can now edit it and then apply it by running ${chalk.greenBright(
          getCommandWithExecutor('prisma migrate --early-access-feature'),
        )} again.`
      } else {
        return `\nNo migration was created. Your Prisma schema and database are already in sync.\n`
      }
    }

    const diagnoseResult = await migrate.diagnoseMigrationHistory()
    debug({ diagnoseResult })

    let isResetNeeded = false
    const hasFailedMigrations = diagnoseResult.failedMigrationNames.length > 0
    const hasEditedMigrations = diagnoseResult.editedMigrationNames.length > 0

    // if failed migration(s) or edited migration(s) print and got to reset
    if (hasFailedMigrations || hasEditedMigrations) {
      isResetNeeded = true

      if (hasFailedMigrations) {
        // migration(s), usually one, that failed to apply the the database (which may have data)
        console.info(
          `The following migration(s) failed to apply:\n- ${diagnoseResult.failedMigrationNames.join(
            '\n- ',
          )}\n`,
        )
      }

      if (hasEditedMigrations) {
        // migration(s) that were edited since they were applied to the db.
        console.info(
          `The following migration(s) were edited after they were applied:\n- ${diagnoseResult.editedMigrationNames.join(
            '\n- ',
          )}\n`,
        )
      }
    } else {
      debug({ drift: diagnoseResult.drift })
      debug({ history: diagnoseResult.history })

      if (diagnoseResult.drift) {
        // TODO it seems this condition is never true
        if (diagnoseResult.drift?.diagnostic === 'migrationFailedToApply') {
          // Migration has a problem (failed to cleanly apply to a temporary database) and
          // needs to be fixed or the database has a problem (example: incorrect version, missing extension)
          throw new Error(
            `A migration failed while applying it to the shadow database.\nFix the migration before applying it again.\n\n${diagnoseResult.drift.error})`,
          )
        } else if (diagnoseResult.drift.diagnostic === 'driftDetected') {
          if (diagnoseResult.hasMigrationsTable === false) {
            const confirmDbPushUsed = await this.confirmDbPushUsed()
            if (confirmDbPushUsed) {
              const introspectEngine = new IntrospectionEngine({
                cwd: path.dirname(schemaPath),
              })

              const introspectResult = await introspectEngine.introspect(
                migrate.getDatamodel(),
              )
              introspectEngine.stop()

              const createMigrationResult = await migrate.createMigration({
                migrationsDirectoryPath: migrate.migrationsDirectoryPath,
                migrationName: '',
                draft: true,
                prismaSchema: introspectResult.datamodel,
              })

              console.info(
                `Migration "${createMigrationResult.generatedMigrationName!}" created.`,
              )

              await migrate.markMigrationApplied({
                migrationId: createMigrationResult.generatedMigrationName!,
              })

              console.info(
                `Migration "${createMigrationResult.generatedMigrationName!}" marked applied.`,
              )

              const createMigrationOptionalResult = await migrate.createMigration(
                {
                  migrationsDirectoryPath: migrate.migrationsDirectoryPath,
                  migrationName: '',
                  draft: false,
                  prismaSchema: migrate.getDatamodel(),
                },
              )

              if (createMigrationOptionalResult.generatedMigrationName) {
                console.info(
                  `Migration "${createMigrationResult.generatedMigrationName!}" created and applied.`,
                )
              }
              migrate.stop()
              return `Operation successful.`
            } else {
              migrate.stop()
              throw Error(
                'Check init flow with introspect + SQL schema dump (TODO docs)',
              )
            }
          } else {
            // we could try to fix the drift in the future
            isResetNeeded = true
          }
        }
      }

      if (diagnoseResult.history) {
        if (diagnoseResult.history.diagnostic === 'databaseIsBehind') {
          await migrate.applyOnly()
        } else if (
          diagnoseResult.history.diagnostic === 'migrationsDirectoryIsBehind'
        ) {
          isResetNeeded = true
        } else if (diagnoseResult.history.diagnostic === 'historiesDiverge') {
          isResetNeeded = true
          // migration(s) were removed from directory since they were applied to the db.
          console.info(
            `The following migration(s) are applied to the database but missing from the local migrations directory:
- ${diagnoseResult.history.unpersistedMigrationNames.join('\n- ')}\n`,
          )
        }
      }
    }

    if (isResetNeeded) {
      if (!args['--force']) {
        // We use prompts.inject() for testing in our CI
        if (isCi() && Boolean((prompt as any)._injected?.length) === false) {
          throw Error(
            `Use the --force flag to use the migrate command in an unnattended environment like ${chalk.bold.greenBright(
              getCommandWithExecutor(
                'prisma migrate --force --early-access-feature',
              ),
            )}`,
          )
        }

        const confirmedReset = await this.confirmReset(
          await migrate.getDbInfo(),
        )
        if (!confirmedReset) {
          console.info() // empty line
          console.info('Reset cancelled.')
          process.exit(0)
        }
        await migrate.reset()
      }
    }

    const evaluateDataLossResult = await migrate.evaluateDataLoss()

    // throw error
    handleUnexecutableSteps(evaluateDataLossResult.unexecutableSteps)
    // log warnings and prompt user to continue if needed
    const userCancelled = await handleWarnings(
      evaluateDataLossResult.warnings,
      args['--force'],
    )
    if (userCancelled) {
      migrate.stop()
      return `Migration cancelled.`
    }

    let migrationName: undefined | string = undefined
    if (evaluateDataLossResult.migrationSteps.length > 0) {
      const getMigrationNameResult = await getMigrationName(args['--name'])
      if (getMigrationNameResult.userCancelled) {
        migrate.stop()
        return getMigrationNameResult.userCancelled
      } else {
        migrationName = getMigrationNameResult.name
      }
    }

    const migrationIds = await migrate.createAndApply({
      name: migrationName,
    })
    migrate.stop()

    if (migrationIds.length === 0) {
      console.info(
        `\n${chalk.green(
          'Everything is already in sync',
        )} - Prisma Migrate didn't find any schema changes or unapplied migrations.`,
      )
    } else {
      console.info(
        `\nPrisma Migrate applied the following migration(s):\n\n${chalk(
          printFilesFromMigrationIds('migrations', migrationIds, {
            'migration.sql': '',
          }),
        )}`,
      )

      // Run if not skipped
      if (!process.env.MIGRATE_SKIP_GENERATE && !args['--skip-generate']) {
        await migrate.tryToRunGenerate()
      }
    }

    return ``
  }

  private async confirmReset({
    schemaWord,
    dbType,
    dbName,
    dbLocation,
  }): Promise<boolean> {
    const confirmation = await prompt({
      type: 'confirm',
      name: 'value',
      message: `We need to reset the ${dbType} ${schemaWord} "${dbName}" at "${dbLocation}". ${chalk.red(
        'All data will be lost',
      )}.\nDo you want to continue?`,
    })

    return confirmation.value
  }

  private async confirmDbPushUsed(): Promise<boolean> {
    const confirmation = await prompt({
      type: 'confirm',
      name: 'value',
      message: `Did you use ${chalk.green('prisma db push')}?`,
    })

    return confirmation.value
  }

  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(
        `\n${chalk.bold.red(`!`)} ${error}\n${MigrateCommand.help}`,
      )
    }
    return MigrateCommand.help
  }
}
