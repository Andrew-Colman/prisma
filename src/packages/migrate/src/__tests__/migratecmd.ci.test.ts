process.env.GITHUB_ACTIONS = '1'

import fs from 'fs-jetpack'
import { MigrateCommand } from '../commands/MigrateCommand'
import { consoleContext, Context } from './__helpers__/context'
import { tearDownMysql } from '../utils/setupMysql'
import { SetupParams, tearDownPostgres } from '../utils/setupPostgres'

const ctx = Context.new().add(consoleContext()).assemble()

describe('common', () => {
  it('migrate in folder with schema only no migrations directory should fail', async () => {
    ctx.fixture('schema-only')
    const result = MigrateCommand.new().parse(['--experimental'])
    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(
      `You need to initialize the migrations by running prisma migrate init --experimental.`,
    )
    expect(ctx.mocked['console.log'].mock.calls.length).toEqual(0)
    expect(ctx.mocked['console.error'].mock.calls.length).toEqual(0)
    expect(
      ctx.mocked['console.info'].mock.calls.join('\n'),
    ).toMatchInlineSnapshot(`Prisma Schema loaded from prisma/schema.prisma`)
  })

  it('migrate should fail if no schema file', async () => {
    ctx.fixture('empty')
    const result = MigrateCommand.new().parse(['--experimental'])
    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(`
                      Could not find a schema.prisma file that is required for this command.
                      You can either provide it with --schema, set it as \`prisma.schema\` in your package.json or put it into the default location ./prisma/schema.prisma https://pris.ly/d/prisma-schema-location
                  `)
  })
})

describe('sqlite', () => {
  it('migrate first migration after init - empty schema', async () => {
    ctx.fixture('initialized-sqlite')
    const result = MigrateCommand.new().parse([
      '--schema=./prisma/empty.prisma',
      '--experimental',
    ])
    await expect(result).resolves.toMatchInlineSnapshot(`

                        Everything is already in sync, Prisma Migrate didn't find any schema changes or unapplied migrations.

                    `)

    expect(ctx.mocked['console.info'].mock.calls.join('\n'))
      .toMatchInlineSnapshot(`
      Prisma Schema loaded from prisma/empty.prisma

      SQLite database dev.db created at file:dev.db

    `)
    expect(ctx.mocked['console.log'].mock.calls.length).toEqual(0)
    expect(ctx.mocked['console.error'].mock.calls.length).toEqual(0)
  })

  it('migrate first migration after init', async () => {
    ctx.fixture('initialized-sqlite')
    const result = MigrateCommand.new().parse([
      '--name=first',
      '--experimental',
    ])
    await expect(result).resolves.toMatchInlineSnapshot(`

                                    Prisma Migrate created and applied the migration 20201231000000_first in

                                    migrations/
                                      └─ 20201231000000_first/
                                        └─ migration.sql

                              `)

    expect(ctx.mocked['console.info'].mock.calls.join('\n'))
      .toMatchInlineSnapshot(`
      Prisma Schema loaded from prisma/schema.prisma

      SQLite database dev.db created at file:dev.db

    `)
    expect(ctx.mocked['console.log'].mock.calls.length).toEqual(0)
    expect(ctx.mocked['console.error'].mock.calls.length).toEqual(0)
  })

  it('migrate first migration after init --force', async () => {
    ctx.fixture('initialized-sqlite')
    const result = MigrateCommand.new().parse([
      '--name=first',
      '--force',
      '--experimental',
    ])
    await expect(result).resolves.toMatchInlineSnapshot(`

                                    Prisma Migrate created and applied the migration 20201231000000_first in

                                    migrations/
                                      └─ 20201231000000_first/
                                        └─ migration.sql

                              `)

    expect(ctx.mocked['console.info'].mock.calls.join('\n'))
      .toMatchInlineSnapshot(`
      Prisma Schema loaded from prisma/schema.prisma

      SQLite database dev.db created at file:dev.db

    `)
    expect(ctx.mocked['console.log'].mock.calls.length).toEqual(0)
    expect(ctx.mocked['console.error'].mock.calls.length).toEqual(0)
  })

  it('create draft migration and apply', async () => {
    ctx.fixture('initialized-sqlite')
    const draftResult = MigrateCommand.new().parse([
      '--draft',
      '--name=first',
      '--experimental',
    ])
    await expect(draftResult).resolves.toMatchInlineSnapshot(`

                                                            Prisma Migrate created a draft migration 20201231000000_first

                                                            You can now edit it and then apply it by running prisma migrate --experimental again.
                                                  `)

    const applyResult = MigrateCommand.new().parse(['--experimental'])
    console.debug('hello', await applyResult)

    await expect(applyResult).resolves.toMatchInlineSnapshot(`

                                                Prisma Migrate created and applied the migration 20201231000000_first in

                                                migrations/
                                                  └─ 20201231000000_first/
                                                    └─ migration.sql

                                        `)

    expect(
      (fs.list('prisma/migrations')?.length || 0) > 0,
    ).toMatchInlineSnapshot(`true`)
    expect(fs.exists('prisma/dev.db')).toEqual('file')

    expect(ctx.mocked['console.info'].mock.calls.join('\n'))
      .toMatchInlineSnapshot(`
      Prisma Schema loaded from prisma/schema.prisma

      SQLite database dev.db created at file:dev.db

      Prisma Schema loaded from prisma/schema.prisma
    `)
    expect(ctx.mocked['console.log'].mock.calls.length).toEqual(0)
    expect(ctx.mocked['console.error'].mock.calls).toMatchInlineSnapshot(
      `Array []`,
    )
  })
})

describe('posgresql', () => {
  const SetupParams: SetupParams = {
    connectionString:
      process.env.TEST_POSTGRES_URI || 'postgres://localhost:5432/prisma-dev',
    dirname: './fixtures',
  }

  beforeEach(async () => {
    await tearDownPostgres(SetupParams).catch((e) => {
      console.debug({ e })
    })
  })

  afterAll(async () => {
    await tearDownPostgres(SetupParams).catch((e) => {
      console.log(e)
    })
  })

  it('migrate first migration after init - empty schema', async () => {
    ctx.fixture('initialized-postgresql')
    const result = MigrateCommand.new().parse([
      '--schema=./prisma/empty.prisma',
      '--experimental',
    ])
    await expect(result).resolves.toMatchInlineSnapshot(`

            Everything is already in sync, Prisma Migrate didn't find any schema changes or unapplied migrations.

          `)

    expect(ctx.mocked['console.info'].mock.calls.join('\n'))
      .toMatchInlineSnapshot(`
      Prisma Schema loaded from prisma/empty.prisma

      PostgreSQL database tests created at localhost:5432

    `)
    expect(ctx.mocked['console.log'].mock.calls.length).toEqual(0)
    expect(ctx.mocked['console.error'].mock.calls.length).toEqual(0)
  })

  it('migrate first migration after init', async () => {
    ctx.fixture('initialized-postgresql')
    const result = MigrateCommand.new().parse([
      '--name=first',
      '--experimental',
    ])
    await expect(result).resolves.toMatchInlineSnapshot(`

                                    Prisma Migrate created and applied the migration 20201231000000_first in

                                    migrations/
                                      └─ 20201231000000_first/
                                        └─ migration.sql

                              `)

    expect(ctx.mocked['console.info'].mock.calls.join('\n'))
      .toMatchInlineSnapshot(`
      Prisma Schema loaded from prisma/schema.prisma

      PostgreSQL database tests created at localhost:5432

    `)
    expect(ctx.mocked['console.log'].mock.calls.length).toEqual(0)
    expect(ctx.mocked['console.error'].mock.calls.length).toEqual(0)
  })

  it('migrate first migration after init --force', async () => {
    ctx.fixture('initialized-postgresql')
    const result = MigrateCommand.new().parse([
      '--name=first',
      '--force',
      '--experimental',
    ])
    await expect(result).resolves.toMatchInlineSnapshot(`

                                    Prisma Migrate created and applied the migration 20201231000000_first in

                                    migrations/
                                      └─ 20201231000000_first/
                                        └─ migration.sql

                              `)

    expect(ctx.mocked['console.info'].mock.calls.join('\n'))
      .toMatchInlineSnapshot(`
      Prisma Schema loaded from prisma/schema.prisma

      PostgreSQL database tests created at localhost:5432

    `)
    expect(ctx.mocked['console.log'].mock.calls.length).toEqual(0)
    expect(ctx.mocked['console.error'].mock.calls.length).toEqual(0)
  })

  it('create draft migration and apply', async () => {
    ctx.fixture('initialized-postgresql')
    const draftResult = MigrateCommand.new().parse([
      '--draft',
      '--name=first',
      '--experimental',
    ])
    await expect(draftResult).resolves.toMatchInlineSnapshot(`

            Prisma Migrate created a draft migration 20201231000000_first

            You can now edit it and then apply it by running prisma migrate --experimental again.
          `)

    const applyResult = MigrateCommand.new().parse(['--experimental'])
    await expect(applyResult).resolves.toMatchInlineSnapshot(`

            Prisma Migrate created and applied the migration 20201231000000_first in

            migrations/
              └─ 20201231000000_first/
                └─ migration.sql

          `)

    expect(ctx.mocked['console.info'].mock.calls.join('\n'))
      .toMatchInlineSnapshot(`
      Prisma Schema loaded from prisma/schema.prisma

      PostgreSQL database tests created at localhost:5432

      Prisma Schema loaded from prisma/schema.prisma
    `)
    expect(ctx.mocked['console.log'].mock.calls.length).toEqual(0)
    expect(ctx.mocked['console.error'].mock.calls).toMatchInlineSnapshot(
      `Array []`,
    )
  })

  it('existingdb: migrate first migration after init', async () => {
    ctx.fixture('initialized-postgresql')
    const result = MigrateCommand.new().parse([
      '--name=first',
      '--experimental',
    ])
    await expect(result).resolves.toMatchInlineSnapshot(`

            Prisma Migrate created and applied the migration 20201231000000_first in

            migrations/
              └─ 20201231000000_first/
                └─ migration.sql

          `)

    expect(ctx.mocked['console.info'].mock.calls.join('\n'))
      .toMatchInlineSnapshot(`
      Prisma Schema loaded from prisma/schema.prisma

      PostgreSQL database tests created at localhost:5432

    `)
    expect(ctx.mocked['console.log'].mock.calls.length).toEqual(0)
    expect(ctx.mocked['console.error'].mock.calls.length).toEqual(0)
  })
})

describe.skip('mysql', () => {
  const SetupParams: SetupParams = {
    connectionString: `${
      process.env.TEST_MYSQL_URI || 'mysql://prisma:prisma@localhost:3306/tests'
    }`,
    dirname: __dirname,
  }

  beforeEach(async () => {
    await tearDownMysql(SetupParams).catch((e) => {
      console.debug({ e })
    })
  })

  afterAll(async () => {
    await tearDownMysql(SetupParams).catch((e) => {
      console.log({ e })
    })
  })
})
