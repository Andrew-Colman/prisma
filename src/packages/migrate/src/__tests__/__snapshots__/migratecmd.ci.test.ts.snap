// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common migrate in folder with schema only no migrations directory should fail 2`] = `Array []`;

exports[`common migrate in folder with schema only no migrations directory should fail 3`] = `Array []`;

exports[`postgresql create draft migration and apply 1`] = `

Prisma Migrate created a draft migration 20201231000000_first

You can now edit it and then apply it by running prisma migrate --experimental again.
`;

exports[`postgresql create draft migration and apply 2`] = ``;

exports[`postgresql create draft migration and apply 4`] = `Array []`;

exports[`postgresql create draft migration and apply 5`] = `Array []`;

exports[`postgresql existingdb: migrate first migration after init 1`] = ``;

exports[`postgresql existingdb: migrate first migration after init 3`] = `Array []`;

exports[`postgresql existingdb: migrate first migration after init 4`] = `Array []`;

exports[`postgresql migrate first migration after init - empty schema 1`] = ``;

exports[`postgresql migrate first migration after init - empty schema 3`] = `Array []`;

exports[`postgresql migrate first migration after init - empty schema 4`] = `Array []`;

exports[`postgresql migrate first migration after init --force 1`] = ``;

exports[`postgresql migrate first migration after init --force 3`] = `Array []`;

exports[`postgresql migrate first migration after init --force 4`] = `Array []`;

exports[`postgresql migrate first migration after init 1`] = ``;

exports[`postgresql migrate first migration after init 3`] = `Array []`;

exports[`postgresql migrate first migration after init 4`] = `Array []`;

exports[`sqlite create draft migration and apply 1`] = `

Prisma Migrate created a draft migration 20201231000000_first

You can now edit it and then apply it by running prisma migrate --experimental again.
`;

exports[`sqlite create draft migration and apply 2`] = ``;

exports[`sqlite create draft migration and apply 5`] = ``;

exports[`sqlite create draft migration and apply 6`] = ``;

exports[`sqlite migrate first migration after init - empty schema 1`] = ``;

exports[`sqlite migrate first migration after init - empty schema 3`] = `Array []`;

exports[`sqlite migrate first migration after init - empty schema 4`] = `Array []`;

exports[`sqlite migrate first migration after init --force 1`] = ``;

exports[`sqlite migrate first migration after init --force 3`] = ``;

exports[`sqlite migrate first migration after init --force 4`] = ``;

exports[`sqlite migrate first migration after init 1`] = ``;

exports[`sqlite migrate first migration after init 3`] = `Array []`;

exports[`sqlite migrate first migration after init 4`] = `Array []`;
